pypit/par/parset.py

```diff
@@ -0,0 +1,649 @@
+            raise ValueError('Input value for {0} invalid: {1}.\nOptions are: {2}'.format(
+                                                                    key, value, self.options[key]))
+        if self.dtype[key] is not None \
+                and not any([ isinstance(value, d) for d in self.dtype[key]]):
+            raise TypeError('Input value for {0} has incorrect type: {1}.'.format(key, value) +
+                            '\nValid types are: {0}'.format(self.dtype[key]))
+
+        if self.can_call[key] and not callable(value):
+            raise TypeError('{0} is not a callable object.'.format(value))
+
+        self.data[key] = value
+
+
+    def __len__(self):
+        """Return the number of parameters."""
+        return self.npar
+        
+
+    def __iter__(self):
+        """Return an iterable to the parameter values."""
+        return iter(self.data.values())
+
+
+    def __repr__(self):
+        """Return a string representation of the parameters."""
+        return self._output_string(header=self.cfg_section)
+
+    
+    def _output_string(self, header=None):
+        additional_par_strings = []
+        data_table = numpy.empty((self.npar+1, 5), dtype=object)
+        data_table[0,:] = ['Parameter', 'Value', 'Default', 'Type', 'Callable']
+        for i, k in enumerate(self.keys()):
+            data_table[i+1,0] = k
+            if isinstance(self.data[k], ParSet):
+                _header = k if header is None else '{0}:{1}'.format(header, k)
+                additional_par_strings += [ self.data[k]._output_string(header=_header) ]
+                data_table[i+1,1] = 'see below'
+                data_table[i+1,2] = 'see below'
+            else:
+                data_table[i+1,1] = ParSet._data_string(self.data[k])
+                data_table[i+1,2] = ParSet._data_string(self.default[k])
+            data_table[i+1,3] = ', '.join([t.__name__ for t in self.dtype[k]])
+            data_table[i+1,4] = self.can_call[k].__repr__()
+
+        output = [ParSet._data_table_string(data_table)]
+        if header is not None:
+            output = [header] + output
+        if len(additional_par_strings) > 0:
+            output += additional_par_strings
+        return '\n'.join(output)
+
+
+    @staticmethod
+    def _data_table_string(data_table):
+        nrows, ncols = data_table.shape
+        col_width = [ numpy.amax([ len(dij) for dij in dj]) for dj in data_table.T ]
+        row_string = ['']*(nrows+1)
+        # Heading row
+        row_string[0] = '  '.join([ data_table[0,j].rjust(col_width[j]) for j in range(ncols)])
+        # Delimiter
+        row_string[1] = '-'*len(row_string[0])
+        for i in range(2,nrows+1):
+            row_string[i] = '  '.join([ data_table[i-1,j].rjust(col_width[j]) 
+                                                                        for j in range(ncols)])
+        return '\n'.join(row_string)+'\n'
+
+
+    @staticmethod
+    def _data_string(data):
```

Comment:
Short doc strings for these methods too.
thanks.

Reply:
Done.